src/coreclr/jit/gentree.cpp

```diff
@@ -14397,9 +14397,10 @@ GenTree* Compiler::gtFoldExprSpecial(GenTree* tree)
             // Optimize boxed value classes; these are always false.  This IL is
             // generated when a generic value is tested against null:
             //     <T> ... foo(T x) { ... if ((object)x == null) ...
-            if ((val == 0) && op->IsBoxedValue())
+            // Also fold checks against known non-null data like static readonlys
+            if ((val == 0) && (op->IsBoxedValue() || !fgAddrCouldBeNull(op)))
```

Comment:
why IsBoxedValue is needed here? shouldn't `fgAddrCouldBeNull` return false for it?

Reply:
Not sure, haven't checked if it does.

Reply:
Okay I checked now and `fgAddrCouldBeNull` returns true for it, it doesn't handle boxes at all.

Reply:
then add to `fgAddrCouldBeNull` ? might catch some other cases as well